src/rules/valid-title.ts

```diff
@@ -0,0 +1,73 @@
+import {
+  createRule,
+  getAccessorValue,
+  getNodeName,
+  isDescribe,
+  isStringNode,
+  isTestCase,
+} from './utils';
+
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+const trimFXprefix = (word: string) =>
+  ['f', 'x'].includes(word.charAt(0)) ? word.substr(1) : word;
+
+const trimStart = (title: string) => /^\s/.test(title);
+
+const getNodeTitle = (node: TSESTree.CallExpression): string | null => {
+  const [argument] = node.arguments;
+
+  if (!isStringNode(argument) || AST_NODE_TYPES.Literal !== argument.type) {
+    return null;
+  }
+
+  return getAccessorValue(argument);
```

Comment:
Thanks for the explanation 👍 Will spend time to understand the types available 😁

Reply:
No problem - I wrote all of the types and guards, and still get confused.

It's actually why I wrote them so grainaulay: The nature of AST parsing means it's best to not try and do fancy things, and instead layer everything to build on the previous type.

As such, the types are something like:

A `StringLiteral` is a `Literal` whose value is of type String (i.e `'hello world'`).
A `TemplateLiteral` is a Template Literal, that has only one quasis (i.e `` `hello world` ``)
  * `quasis` is a property on `TemplateLiteral`s, that is an array of `TemplateElement`s.
    * No idea why it's `quasis` - it's called that in the spec, but I can't find a definition of the word.
  * In theory this should mean only "simple" template literals that don't have any expressions, as a `TemplateElement` sort of represents the "stringy" bits of a template literal i.e `` `hello ${world}` `` has 2 `TemplateElement`s, the first being `hello ${` (w/ a value of `hello `) and the second being `}` (w/ a value of `''`).
  * In practice, this does work, but I think it might be better to check the length of `expression`, to see if it's 0
    * I know `no-identical-title` checks both `isStringNode` & the expression length if it's a template, and fails if I remove the expression length check, which iirc means `isStringNode` isn't doing its job right.

A `StringNode` is a `node` that's either a `StringLiteral` OR a `TemplateLiteral`.

-----

That ends the "string" side of nodes. You then have `Identifiers`, which are pretty much what they say on the tin: they're something that "identifies" as something - best way to think of them is as plaintext "names" or "words" - anything else is another expression, i.e `[]` is an `ArrayExpression`.

Identifiers are what you use to access properties, methods, fields, etc via dot notation. They're also how variables are used. 

A `KnownIdentifier` is an `Identifier` whose name we "known"
  * This is actually a bit of a cheeky one - when we say "known", the "least known" `Identifier` has a `name` of type `string`; that's it. Which the case for *all* `Identifier`s anyway, and is the default. The name was more about reducing confusion between *our* generically typed `Identifier` and `TSESTree.Identifier`, of which ours extends.

--------
An `AccessorNode` is then either a `StringNode` OR a `KnownIdentifier`.

The whole backstory behind `AccessorNode` is that you can access things in two ways in JS: Dot notation, and Bracket notation.

Dot accessors are `Identifiers` - nice and easy. Bracket accessors however are done using "a string" in square brackets; so they're `StringNode`s (b/c they could be string literals or template expressions - they can also be `Identifier`s themselves, b/c variables but we don't support that b/c we're not trying to be TS 😉).

As such, we have `isSupportedAccessor`, which guards on if a `node` is an `AccessorNode` that we "support". Then for when we want to get the value of an `AccessorNode`, we have `getAccessorValue`, which normalises getting the value of the `node` you're working w/.

------

And there you go, that's pretty much it in a nutshell 😄 

While on the surface it can look very complex, it's actually pretty simple to the point of being elegant (IMO at least 😉) - it's done as just a composition chain, which means you just have to follow the type chain down each step to get an idea of what the previous step is made up of.

In saying that, once you throw the actual AST tree into the mix, it's *very* easy to get confused & mentally tongue tied. The above is covering only the property accessor side; theres also the `expect`/modifier/matcher utils, which are also hopefully as elegant, but still their own kettle of fish.

Finally, I'm meant to be writing a whole new set of parsers/guards for the jest globals (`describe`, `test`, `it`, etc), but that's slow going for now (plus what we have is working, so it's less needed).

-----

This is pretty much the best write up I've done on this so far, and something I've been meaning to do for a while. Cheers!

Reply:
Wow, that's so much of information, Thank you very much for taking time to write this up 😊

Reply:
No problem! I've tried to write this up a few times before, but didn't get very far, so when I started writing this comment I just let things snowball so that as much info as possible is on paper :joy: